File: app/src/main/java/com/kaltura/kflow/presentation/productPrice/ProductPriceListAdapter.java
package com.kaltura.kflow.presentation.productPrice;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.recyclerview.widget.RecyclerView;

import com.kaltura.client.types.PpvPrice;
import com.kaltura.client.types.ProductPrice;
import com.kaltura.client.types.SubscriptionPrice;
import com.kaltura.kflow.R;

import java.util.ArrayList;

/**
 * Created by alex_lytvynenko on 30.11.2018.
 */
public class ProductPriceListAdapter extends RecyclerView.Adapter<ProductPriceListAdapter.MyViewHolder> {
    private ArrayList<ProductPrice> mProductPrices = new ArrayList<>();
    private ProductPriceListAdapter.OnProductPriceClickListener mListener;

    ProductPriceListAdapter(ArrayList<ProductPrice> assets, ProductPriceListAdapter.OnProductPriceClickListener listener) {
        mProductPrices.addAll(assets);
        mListener = listener;
    }

    @NonNull
    @Override
    public ProductPriceListAdapter.MyViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ProductPriceListAdapter.MyViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.item_product_price, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull ProductPriceListAdapter.MyViewHolder holder, int position) {
        holder.bind(mProductPrices.get(position), mListener);
    }

    @Override
    public int getItemCount() {
        return mProductPrices.size();
    }

    public interface OnProductPriceClickListener {
        void onSubscriptionPriceClicked(SubscriptionPrice subscriptionPrice);
    }

    class MyViewHolder extends RecyclerView.ViewHolder {

        private AppCompatTextView mType;
        private AppCompatTextView mPrice;

        MyViewHolder(View v) {
            super(v);
            mType = v.findViewById(R.id.product_price_type);
            mPrice = v.findViewById(R.id.price);
        }

        void bind(final ProductPrice productPrice, final ProductPriceListAdapter.OnProductPriceClickListener clickListener) {
            mType.setText("Product Type: " + productPrice.getProductType().getValue());
            mPrice.setText("Price: " + productPrice.getPrice().getCurrencySign() + productPrice.getPrice().getAmount());
            if (productPrice instanceof SubscriptionPrice) {
                itemView.setOnClickListener(view -> clickListener.onSubscriptionPriceClicked((SubscriptionPrice) productPrice));
            } else if (productPrice instanceof PpvPrice) {
                itemView.setOnClickListener(null);
            }
        }
    }
}

File: app/src/main/java/com/kaltura/kflow/presentation/productPrice/ProductPriceListAdapter.kt
package com.kaltura.kflow.presentation.productPrice

import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.kaltura.client.types.PpvPrice
import com.kaltura.client.types.ProductPrice
import com.kaltura.client.types.SubscriptionPrice
import com.kaltura.kflow.R
import com.kaltura.kflow.presentation.extension.inflate
import kotlinx.android.extensions.LayoutContainer
import kotlinx.android.synthetic.main.item_product_price.*
import java.util.*

/**
 * Created by alex_lytvynenko on 30.11.2018.
 */
class ProductPriceListAdapter(private val productPrices: ArrayList<ProductPrice>) : RecyclerView.Adapter<ProductPriceListAdapter.MyViewHolder>() {

    var onSubscriptionPriceClickListener: (subscriptionPrice: SubscriptionPrice) -> Unit = {}

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) = MyViewHolder(parent.inflate(R.layout.item_product_price))

    override fun onBindViewHolder(holder: MyViewHolder, position: Int) = holder.bind(productPrices[position])

    override fun getItemCount() = productPrices.size

    inner class MyViewHolder(override val containerView: View) : RecyclerView.ViewHolder(containerView), LayoutContainer {

        fun bind(productPrice: ProductPrice) {
            productPriceType.text = "Product Type: ${productPrice.productType.value}"
            price.text = "Actual price: ${productPrice.fullPrice.currencySign}${productPrice.fullPrice.amount}"
            discountPrice.text = "Discount price: ${productPrice.price.currencySign}${productPrice.price.amount}"
            if (productPrice is SubscriptionPrice) {
                itemView.setOnClickListener { onSubscriptionPriceClickListener(productPrice) }
            } else if (productPrice is PpvPrice) {
                itemView.setOnClickListener(null)
            }
        }
    }
}